Migrate shared metrics into component base

logicalhan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:

 /kind feature


**What this PR does / why we need it**:

Moves shared adaptors for client-go metrics into comp-base. These use the internal wrapped registries and our wrapping metric framework.



```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:

-->
```docs
- [KEP - Metrics Stability Migration]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190605-metrics-stability-migration.md
- [KEP - Metrics Stability Framework]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md
```

logicalhan commented:
/sig instrumentation
/cc @brancz @serathius

logicalhan commented:
/remove-sig api-machinery cluster-lifecycle

logicalhan commented:
/priority important-soon
/kind feature

brancz commented:
This generally looks good. I'll be happy to get rid of those deprecated workqueue metrics :)

sttts commented on staging/src/k8s.io/component-base/go.mod:
```diff
@@ -13,16 +13,20 @@ require (
 	github.com/spf13/pflag v1.0.3
 	github.com/stretchr/testify v1.3.0
 	k8s.io/apimachinery v0.0.0
+	k8s.io/client-go v0.0.0
```
there is a rules file in staging/ and some import restriction file. Both need to be changed to include/allow client-go here.

logicalhan replied:
done.

sttts commented on staging/src/k8s.io/component-base/metrics/clientgo/leaderelection/metrics.go:
```diff
@@ -0,0 +1,56 @@
+limitations under the License.
+*/
+
+package leaderelection
```
would expect prometheus in the package path.

Am not sure about `clientgo/`. I tend to prefer `client/` only, but am not even that we need the `clientgo/` at all, at least for leader election.

logicalhan replied:
appended prometheus to directory path.

logicalhan commented:
/retest

logicalhan commented:
/assign @sttts

brancz commented:
I don't think I can approve any of this, but lgtm :)

/lgtm
/approve

logicalhan commented:
/retest

brancz commented:
/retest

sttts commented on staging/publishing/import-restrictions.yaml:
```diff
@@ -56,6 +56,8 @@
   - k8s.io/apimachinery
   - k8s.io/component-base
+  - k8s.io/client-go
+  - k8s.io/client-go/tools/
```
is this necessary?

logicalhan replied:
Yes, as I understand it 'k8s.io/client-go/tools/' is an explicitly restricted directory.

sttts commented on staging/src/k8s.io/component-base/metrics/shared/restclient/BUILD:
```diff
@@ -0,0 +1,32 @@
+package(default_visibility = ["//visibility:public"])
```
what does "shared" mean in the package path? It is a shared library obviously. But do we need that in the path?

logicalhan replied:
Sorry that was out of date.

logicalhan commented:
/retest

Commit:
move version logic directly into the component-base, to simplify legacyregistry logic

add version to vendor

move version into component-base top-level

update bazel build files

fix import

logicalhan commented:
/retest

logicalhan commented:
/retest

logicalhan commented:
/retest

sttts commented on staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go:
```diff
@@ -17,112 +17,53 @@ limitations under the License.
+//
+// Deprecated: Please note the issues described in the doc comment of
+// InstrumentHandler. You might want to consider using promhttp.Handler instead.
+func Handler() http.Handler {
```
if it is deprecated, why do we change the signature?

logicalhan replied:
This method intends to replace the method calls to the original prometheus global Handler (https://github.com/prometheus/client_golang/blob/7ed96b33bd86448bfd5915bf4f0f96b7eb826108/prometheus/http.go#L47-L55), which has the same method signature as this method and is also similarly deprecated (but in use in our codebase).

logicalhan replied:
Specifically, we require this for backwards compatibility reasons. The InstrumentedHandler effectively adds a series of summary metrics which are, by default, added to the global prometheus registry. This method allows us to maintain parity with the metrics that the native global prometheus registry.

sttts commented on staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go:
```diff
@@ -17,112 +17,53 @@ limitations under the License.
+// bypasses the metric stability framework
+//
+// Deprecated
+func RawMustRegister(cs ...prometheus.Collector) {
```
why do we add a deprecated func?

logicalhan replied:
same as below.

sttts commented on staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go:
```diff
@@ -17,112 +17,53 @@ limitations under the License.
+// bypasses the metric stability framework
+//
+// Deprecated
+func RawRegister(c prometheus.Collector) error {
```
why do we add a deprecated func?

logicalhan replied:
Because we eventually do not want to be registering native prometheus Collectors at all, since they cannot be annotated with stability information. We require them at the moment due to the pervasive use of custom prometheus collectors in the kubelet, though we do not want to encourage or increase usage of this method going forward. @serathius is working on additional static analysis checks to aid that effort.

s-urbaniak replied:
good call for the immediate deprecation :+1: Can we add some explicit sentence in the godoc that this is not meant to be used for new code and point them to an alternative, i.e. native metrics? deprecated methods do tend to be used in newly added code.

sttts commented on staging/src/k8s.io/component-base/metrics/registry.go:
```diff
@@ -139,11 +140,22 @@ func (kr *kubeRegistry) Gather() ([]*dto.MetricFamily, error) {
+
+// NewKubeRegistry creates a new vanilla Registry without any Collectors
+// pre-registered.
+func NewKubeRegistry() KubeRegistry {
```
this breaks old code, no?

logicalhan replied:
Nothing is migrated yet.

sttts commented on staging/src/k8s.io/component-base/metrics/version/version.go:
```diff
@@ -0,0 +1,42 @@
+
+// Get returns the overall codebase version. It's for detecting
+// what code a binary was built from.
+func Get() apimachineryversion.Info {
```
where is this called?

logicalhan replied:
https://github.com/kubernetes/kubernetes/blob/6c72548265d96dc40d8f0e8592997aaa041575ab/staging/src/k8s.io/component-base/metrics/version.go#L34

sttts replied:
don't we have the same already elsewhere? E.g. in client-go?

sttts commented on staging/src/k8s.io/component-base/metrics/prometheus/clientgo/BUILD:
```diff
@@ -0,0 +1,31 @@
+package(default_visibility = ["//visibility:public"])
```
It's not a move (as in the commit message). It's a duplication.

logicalhan replied:
okay, i will amend the commit message.

sttts commented on vendor/modules.txt:
```diff
@@ -1575,6 +1575,7 @@ k8s.io/component-base/featuregate
 k8s.io/component-base/featuregate/testing
 k8s.io/component-base/logs
 k8s.io/component-base/metrics
+k8s.io/component-base/metrics/version
```
this change is in the wrong commit

logicalhan replied:
ah yes it is

logicalhan commented:
/retest

sttts commented on staging/src/k8s.io/component-base/metrics/version/version.go:
```diff
@@ -0,0 +1,42 @@
+		GoVersion:    runtime.Version(),
+		Compiler:     runtime.Compiler,
+		Platform:     fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+	}
```
there is no connection to metrics whatsoever, is there?

Commit:
duplicate migrated variants of shared metrics into component-base

logicalhan commented:
/retest

brancz commented:
/lgtm
/approve

liggitt commented on hack/lib/version.sh:
```diff
@@ -161,6 +161,7 @@ kube::version::ldflags() {
       "-X '${KUBE_GO_PACKAGE}/vendor/k8s.io/client-go/pkg/version.${key}=${val}'"
       "-X '${KUBE_GO_PACKAGE}/vendor/k8s.io/kubectl/pkg/version.${key}=${val}'"
```
please open a follow up issue to collapse the kubectl pkg/version and consumers of the k8s.io/kubernetes/pkg/version onto component-base/version

logicalhan replied:
https://github.com/kubernetes/kubernetes/issues/81731

liggitt commented:
/approve

go.mod and dependency changes lgtm. deferring to @brancz on component-base/metrics review.

/hold
would like someone from test-infra to ack the `staging/src/k8s.io/component-base/version/def.bzl` addition, then can unhold
/cc @BenTheElder

logicalhan commented:
test /pull-kubernetes-integration

logicalhan commented:
/test pull-kubernetes-integration

sttts commented on hack/lib/version.sh:
```diff
@@ -161,6 +161,7 @@ kube::version::ldflags() {
       "-X '${KUBE_GO_PACKAGE}/vendor/k8s.io/client-go/pkg/version.${key}=${val}'"
       "-X '${KUBE_GO_PACKAGE}/vendor/k8s.io/kubectl/pkg/version.${key}=${val}'"
       "-X '${KUBE_GO_PACKAGE}/cmd/kubeadm/app/version.${key}=${val}'"
+      "-X '${KUBE_GO_PACKAGE}/vendor/k8s.io/component-base/version.${key}=${val}'"
```
@stealthybox this needs follow-ups in all of our components.

sttts replied:
i.e. use this, not pkg/version

logicalhan replied:
https://github.com/kubernetes/kubernetes/issues/81731

logicalhan replied:
Yeah, we have like a bazillion copies of this thing currently everywhere.

sttts commented:
Am fine with the structure. +1 for test-infra and @brancz for final lgtm.

Commit:
add client-go as a dep of component-base

logicalhan commented:
/cc @fejta

logicalhan commented:
/assign @ehashman

fejta reviewed (commented):
/lgtm

liggitt commented:
/hold cancel

for fejta's bazel ack

piosz commented:
/lgtm
/approve

BenTheElder commented on staging/src/k8s.io/component-base/version/def.bzl:
```diff
@@ -0,0 +1,39 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Implements hack/lib/version.sh's kube::version::ldflags() for Bazel.
```
ACK.

If this is intended to be consumed by many repos, we should probably include usage somewhere?

BenTheElder commented on staging/src/k8s.io/component-base/version/def.bzl:
```diff
@@ -0,0 +1,39 @@
+    ]
+
+    # This should match the list of vars in kube::version::ldflags
+    # It should also match the list of vars set in hack/print-workspace-status.sh.
```
I.E. a user will need this setup, amongst other things..
